File: CSharp/src/Runtime/Scripts/XLog/adapter.cs
// Copyright (c) 2026 FrameworkEase Technologies. All rights reserved.
// Use of this source code is governed by a MIT-style
// license that can be found in the LICENSE file.

using System;
using System.Collections.Generic;

namespace FrameworkEase.Universal.X.Utility
{
    public partial class XLog
    {
        public interface IAdapter
        {
            string Name { get; }
            void Setup();
            void Write(Data data);
            void Flush();
            void Reset();
        }

        public static class Adapter
        {
            internal static readonly object mutex = new();
            internal static readonly List<IAdapter> adapters = new();

            public static void Range(Func<IAdapter, bool> callback)
            {
                if (callback == null) return;
                foreach (var apt in adapters) if (!callback(apt)) break;
            }

            public static bool Setup(params IAdapter[] apt)
            {
                if (apt == null || apt.Length == 0) return false;
                lock (mutex)
                {
                    var ok = true;
                    foreach (var ele in apt)
                    {
                        if (ele == null)
                        {
                            ok = false;
                            continue;
                        }
                        if (adapters.Contains(ele))
                        {
#if UNITY_5_3_OR_NEWER
                            UnityHandler.origin.LogFormat(UnityEngine.LogType.Error, null, $"XLog.Adapter.Setup: duplicated adapter: {ele.Name}.");
#else
                            Console.Error.WriteLine($"XLog.Adapter.Setup: duplicated adapter: {ele.Name}.");
#endif
                            ok = false;
                            continue;
                        }
                        adapters.Add(ele);
                        ele.Setup();
#if UNITY_5_3_OR_NEWER
                        UnityHandler.origin.LogFormat(UnityEngine.LogType.Log, null, $"XLog.Adapter.Setup: registered adapter: {ele.Name}.");
#else
                        Console.WriteLine($"XLog.Adapter.Setup: registered adapter: {ele.Name}.");
#endif
                    }
                    return ok;
                }
            }

#if UNITY_5_3_OR_NEWER
            [UnityEngine.HideInCallstack]
#endif
            public static void Write(Data data)
            {
                if (adapters.Count == 0)
                {
#if UNITY_5_3_OR_NEWER
                    UnityHandler.origin.LogFormat(UnityEngine.LogType.Log, null, "{0}", data.Stringify());
#else
                    Console.WriteLine(data.Stringify());
#endif
                }
                else foreach (var apt in adapters) apt.Write(data);
            }

            public static void Flush()
            {
                lock (mutex)
                {
#if UNITY_5_3_OR_NEWER
                    UnityHandler.origin.LogFormat(UnityEngine.LogType.Log, null, $"XLog.Adapter.Flush: performing flush with {adapters.Count} adapter(s).");
#else
                    Console.WriteLine($"XLog.Adapter.Flush: performing flush with {adapters.Count} adapter(s).");
#endif
                    foreach (var apt in adapters) apt.Flush();
                }
            }

            public static bool Reset(IAdapter apt = null)
            {
                lock (mutex)
                {
                    if (apt == null)
                    {
                        var cnt = adapters.Count;
                        foreach (var ele in adapters) ele.Reset();
                        adapters.Clear();
#if UNITY_5_3_OR_NEWER
                        UnityHandler.origin.LogFormat(UnityEngine.LogType.Log, null, $"XLog.Adapter.Reset: unregistered {cnt} adapter(s).");
#else
                        Console.WriteLine($"XLog.Adapter.Reset: unregistered {cnt} adapter(s).");
#endif
                    }
                    else
                    {
                        var idx = adapters.IndexOf(apt);
                        if (idx < 0) return false;
                        adapters.RemoveAt(idx);
                        apt.Reset();
#if UNITY_5_3_OR_NEWER
                        UnityHandler.origin.LogFormat(UnityEngine.LogType.Log, null, $"XLog.Adapter.Reset: unregistered adapter: {apt.Name}.");
#else
                        Console.WriteLine($"XLog.Adapter.Reset: unregistered adapter: {apt.Name}.");
#endif
                    }
                    return true;
                }
            }
        }
    }
}

File: Go/XLog/adapter.go
// Copyright (c) 2026 FrameworkEase Technologies. All rights reserved.
// Use of this source code is governed by a MIT-style
// license that can be found in the LICENSE file.

package XLog

import (
	"fmt"
	"os"
	"slices"
	"sync"
)

type IAdapter interface {
	Name() string
	Setup()
	Write(data Data)
	Flush()
	Reset()
}

type adapterScope struct {
	mutex    sync.RWMutex
	adapters []IAdapter
}

var Adapter = &adapterScope{}

func (as *adapterScope) Setup(apt ...IAdapter) bool {
	if len(apt) == 0 {
		return false
	}
	as.mutex.Lock()
	defer as.mutex.Unlock()

	ok := true
	for _, ele := range apt {
		if ele == nil {
			ok = false
			continue
		}
		if slices.Contains(as.adapters, ele) {
			fmt.Fprintf(os.Stderr, "XLog.Adapter.Setup: duplicated adapter: %v.\n", ele.Name())
			ok = false
			continue
		}
		as.adapters = append(as.adapters, ele)
		ele.Setup()
		fmt.Printf("XLog.Adapter.Setup: registered adapter: %s.\n", ele.Name())
	}
	return ok
}

func (as *adapterScope) Reset(apt ...IAdapter) bool {
	as.mutex.Lock()
	defer as.mutex.Unlock()

	if len(apt) == 0 {
		cnt := len(as.adapters)
		for _, ele := range as.adapters {
			ele.Reset()
		}
		as.adapters = nil
		fmt.Printf("XLog.Adapter.Reset: unregistered %d adapter(s).\n", cnt)
	} else {
		ele := apt[0]
		i := slices.Index(as.adapters, ele)
		if i < 0 {
			return false
		}
		as.adapters = append(as.adapters[:i], as.adapters[i+1:]...)
		ele.Reset()
		fmt.Printf("XLog.Adapter.Reset: unregistered adapter: %v.\n", ele.Name())
	}
	return true
}

func (as *adapterScope) Range(callback func(apt IAdapter) bool) {
	if callback == nil {
		return
	}
	for _, apt := range as.adapters {
		if !callback(apt) {
			break
		}
	}
}

func (as *adapterScope) Write(data Data) {
	if len(as.adapters) == 0 {
		fmt.Println(data.Stringify())
	} else {
		for _, apt := range as.adapters {
			apt.Write(data)
		}
	}
}

func (as *adapterScope) Flush() {
	fmt.Printf("XLog.Adapter.Flush: performing flush with %d adapter(s).\n", len(as.adapters))
	as.mutex.RLock()
	defer as.mutex.RUnlock()

	for _, apt := range as.adapters {
		apt.Flush()
	}
}
